pkg/metrics/metrics.go

```diff
@@ -0,0 +1,220 @@
+	if opts.lookuper == nil {
+		opts.lookuper = envconfig.OsLookuper()
+	}
+
+	optOut, err := optout.LoadOptOutSettings(ctx, opts.lookuper, appID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load opt out settings: %w", err)
+	}
+
+	// Short Circuit if user opted out of metrics.
+	if optOut.NoMetrics {
+		return &Client{optOut: optOut}, nil
+	}
+
+	// Default to 1 second timeout httpClient.
+	if opts.httpClient == nil {
+		opts.httpClient = &http.Client{Timeout: 1 * time.Second}
+	}
+
+	var c metricsConfig
+	prefixLookuper := envconfig.PrefixLookuper(strings.ToUpper(appID)+"_", opts.lookuper)
+	if err := envconfig.ProcessWith(ctx, &envconfig.Config{
+		Target:   &c,
+		Lookuper: prefixLookuper,
+	}); err != nil {
+		return nil, fmt.Errorf("failed to process envconfig: %w", err)
+	}
+
+	// Use ParseRequestURI over Parse because Parse validation is more loose and will accept
+	// things such as relative paths without a host.
+	if _, err := url.ParseRequestURI(c.ServerURL); err != nil {
+		return nil, fmt.Errorf("failed to parse server url: %w", err)
+	}
+
+	storedID, err := loadInstallID(appID, opts.installIDFileOverride)
+	var installID string
+	if err != nil || storedID == nil {
+		installID, err = generateInstallID()
+		if err != nil {
+			return nil, err
+		}
+
+		if err = storeInstallID(appID, opts.installIDFileOverride, &InstallIDData{
+			InstallID: installID,
+		}); err != nil {
+			logging.FromContext(ctx).DebugContext(ctx, "error storing installID", "error", err.Error())
+		}
+	} else {
+		installID = storedID.InstallID
+	}
+
+	return &Client{
+		appID:      appID,
+		version:    version,
+		installID:  installID,
+		httpClient: opts.httpClient,
+		optOut:     optOut,
+		config:     &c,
+	}, nil
+}
+
+type SendMetricRequest struct {
+	// The ID of the application to check.
+	AppID string `json:"appId"`
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string `json:"version"`
+
+	Metrics map[string]int `json:"metrics"`
+
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+// SendSync sends information about application usage. Noop if metrics
+// are opted out.
+// Accepts a context for cancellation.
+func (c *Client) SendSync(ctx context.Context, metrics map[string]int) error {
+	if c.optOut.NoMetrics {
+		return nil
+	}
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(&SendMetricRequest{
+		AppID:     c.appID,
+		Version:   c.version,
+		Metrics:   metrics,
+		InstallID: c.installID,
+	}); err != nil {
+		return fmt.Errorf("failed to marshal metrics as json: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, fmt.Sprintf(c.config.ServerURL+"/sendMetrics"), &buf)
+	if err != nil {
+		return fmt.Errorf("failed to create http request: %w", err)
+	}
+	req.Header.Set("User-Agent", "Go-http-client/1.1 github.com/abcxyz/abc-updater")
```

Comment:
I don't know the protocol requirements here, so ignore me if this is bad advice. Do we really want to include `Go-http-client/1.1` here? In the future we're likely to update to a new version of the go http client, and we're unlikely to remember to come change this code. Can this part be left out?

Reply:
Yeah, I'll leave out